erpnext/controllers/buying_controller.py

```diff
@@ -99,6 +100,21 @@ def validate_stock_or_nonstock_items(self):
 				for d in tax_for_valuation:
 					d.category = 'Total'
 				msgprint(_('Tax Category has been changed to "Total" because all the Items are non-stock items'))
+	
+	def validate_asset_return(self):
+		if self.doctype not in ['Purchase Receipt', 'Purchase Invoice'] or not self.is_return:
+			return
+
+		purchase_doc_field = 'purchase_receipt' if self.doctype == 'Purchase Receipt' else 'purchase_invoice'
+		not_cancelled_asset = [d.name for d in frappe.db.get_all("Asset", {
+			purchase_doc_field: self.return_against,
+			"docstatus": 1
+		})]
+		if self.is_return and len(not_cancelled_asset):
+			if len(not_cancelled_asset) < 4:
```

Comment:
Why less than 4?

Reply:
Just for better message. If there are less than 4 assets not cancelled then display their names in the error message.